File: electron/openaiChat.js
/**
 * OpenAI 兼容 API 适配（可配置 baseURL）。
 * 不包含模型名称或参数，全部由调用方通过 options 传入。
 */

const OpenAI = require('openai')
const {
  SESSION_TITLE_SYSTEM_PROMPT,
  normalizeSessionTitle,
} = require('./sessionTitle')

function createClient(apiKey, baseURL) {
  const url = typeof baseURL === 'string' ? baseURL.trim() : ''
  if (!url) {
    throw new Error('请填写接口地址')
  }
  return new OpenAI({
    apiKey,
    baseURL: url.replace(/\/+$/, ''),
  })
}

// ─── 非流式对话（带 tools 时用，保证拿到完整 tool_calls 含 id）────────────────
/**
 * @returns {Promise<{ message: object, model: string }>}
 */
async function chatNoStream(apiKey, messages, options = {}, signal) {
  const { model, temperature, thinking, tools, max_tokens, baseURL } = options
  const client = createClient(apiKey, baseURL)
  const params = { model, messages, stream: false }
  if (temperature !== undefined && temperature !== null) params.temperature = temperature
  if (thinking) params.thinking = thinking
  if (max_tokens) params.max_tokens = max_tokens
  if (tools && tools.length > 0) params.tools = tools
  const res = await client.chat.completions.create(params, { signal })
  const message = res.choices?.[0]?.message || {}
  return { message, model: res.model || model }
}

// ─── 流式对话 ────────────────────────────────────────────────────
/**
 * @param {string} apiKey
 * @param {Array} messages
 * @param {{ model: string, temperature?: number, thinking?: object, tools?: array, max_tokens?: number }} options
 * @returns {Promise<{ stream: AsyncIterable, model: string }>}
 */
async function chatStream(apiKey, messages, options = {}, signal) {
  const { model, temperature, thinking, tools, max_tokens, baseURL } = options
  const toolNames = tools?.length ? tools.map((t) => t.function?.name).filter(Boolean).join(', ') : ''
  const tempLog =
    temperature !== undefined && temperature !== null ? `temperature=${temperature}` : 'temperature=(omit)'
  console.log(`[OpenAI] ${model} ${tempLog}${thinking ? ` thinking=${thinking.type}` : ''}${toolNames ? ` tools=${toolNames}` : ''}`)

  const client = createClient(apiKey, baseURL)
  const params = { model, messages, stream: true }
  if (temperature !== undefined && temperature !== null) params.temperature = temperature
  if (thinking) params.thinking = thinking
  if (max_tokens) params.max_tokens = max_tokens
  if (tools && tools.length > 0) params.tools = tools
  const stream = await client.chat.completions.create(params, { signal })
  return { stream, model }
}

// ─── 标题生成（非流式 / temperature=1 / 不用思考扩展字段）────────────────
/**
 * @param {string} apiKey
 * @param {string} text
 * @param {{ model: string, baseURL?: string }} options
 * @returns {Promise<string>} 已 normalize，最多 10 字
 */
async function generateTitle(apiKey, text, options = {}) {
  const { model, baseURL } = options
  const client = createClient(apiKey, baseURL)
  const res = await client.chat.completions.create({
    model,
    messages: [
      { role: 'system', content: SESSION_TITLE_SYSTEM_PROMPT },
      { role: 'user', content: String(text || '').trim() },
    ],
    max_tokens: 32,
    thinking: { type: 'disabled' },
    stream: false,
  })
  const raw = res.choices?.[0]?.message?.content ?? ''
  console.log('[ai-generate-title][openai] 模型返回原文:', String(raw))
  return normalizeSessionTitle(raw)
}

module.exports = { chatStream, chatNoStream, generateTitle }

File: backend/services/openai_chat.py
"""
OpenAI-compatible chat adapter — port of electron/openaiChat.js.

Uses the ``openai`` async Python SDK and exposes streaming / non-streaming
chat plus a title-generation helper.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any, AsyncIterator

from openai import AsyncOpenAI

from utils.session_title import (
    SESSION_TITLE_SYSTEM_PROMPT,
    normalize_session_title,
)
from utils.url import normalize_base_url

logger = logging.getLogger(__name__)


def _create_client(api_key: str, base_url: str | None) -> AsyncOpenAI:
    url = normalize_base_url(base_url)
    if not url:
        raise ValueError("请填写接口地址")
    return AsyncOpenAI(api_key=api_key, base_url=url)


# ── Non-streaming chat ──────────────────────────────────────────

async def chat_no_stream(
    api_key: str,
    messages: list[dict],
    options: dict[str, Any] | None = None,
    signal: asyncio.Event | None = None,
) -> dict[str, Any]:
    """Single-shot completion.  Returns ``{"message": {...}, "model": str}``."""
    opts = options or {}
    model: str = opts.get("model", "")
    temperature = opts.get("temperature")
    thinking = opts.get("thinking")
    tools: list | None = opts.get("tools")
    max_tokens: int | None = opts.get("max_tokens")
    base_url: str | None = opts.get("baseURL")
    top_k: Any = opts.get("top_k")

    client = _create_client(api_key, base_url)

    params: dict[str, Any] = {"model": model, "messages": messages, "stream": False}
    if temperature is not None:
        params["temperature"] = temperature
    if thinking:
        params.setdefault("extra_body", {})["thinking"] = thinking
    if max_tokens:
        params["max_tokens"] = max_tokens
    if tools:
        params["tools"] = tools
    if top_k is not None:
        try:
            tk = int(top_k)
            if tk > 0:
                params.setdefault("extra_body", {})["top_k"] = tk
        except (TypeError, ValueError):
            pass

    res = await client.chat.completions.create(**params)
    choice = res.choices[0] if res.choices else None
    message = choice.message.model_dump() if choice and choice.message else {}
    return {"message": message, "model": res.model or model}


# ── Streaming chat ──────────────────────────────────────────────

async def chat_stream(
    api_key: str,
    messages: list[dict],
    options: dict[str, Any] | None = None,
    signal: asyncio.Event | None = None,
) -> dict[str, Any]:
    """Streaming completion.  Returns ``{"stream": async_generator, "model": str}``."""
    opts = options or {}
    model: str = opts.get("model", "")
    temperature = opts.get("temperature")
    thinking = opts.get("thinking")
    tools: list | None = opts.get("tools")
    max_tokens: int | None = opts.get("max_tokens")
    base_url: str | None = opts.get("baseURL")
    top_k: Any = opts.get("top_k")

    tool_names = (
        ", ".join(t.get("function", {}).get("name", "") for t in tools if t.get("function", {}).get("name"))
        if tools
        else ""
    )
    temp_log = f"temperature={temperature}" if temperature is not None else "temperature=(omit)"
    thinking_log = f" thinking={thinking.get('type')}" if thinking else ""
    tools_log = f" tools={tool_names}" if tool_names else ""
    logger.info("[OpenAI] %s %s%s%s", model, temp_log, thinking_log, tools_log)

    client = _create_client(api_key, base_url)

    params: dict[str, Any] = {"model": model, "messages": messages, "stream": True}
    if temperature is not None:
        params["temperature"] = temperature
    if thinking:
        params.setdefault("extra_body", {})["thinking"] = thinking
    if max_tokens:
        params["max_tokens"] = max_tokens
    if tools:
        params["tools"] = tools
    if top_k is not None:
        try:
            tk = int(top_k)
            if tk > 0:
                params.setdefault("extra_body", {})["top_k"] = tk
        except (TypeError, ValueError):
            pass

    raw_stream = await client.chat.completions.create(**params)

    async def _generate() -> AsyncIterator[dict]:
        async for chunk in raw_stream:
            if signal and signal.is_set():
                break
            yield chunk.model_dump()

    return {"stream": _generate(), "model": model}


# ── Title generation ────────────────────────────────────────────

async def generate_title(
    api_key: str,
    text: str,
    options: dict[str, Any] | None = None,
) -> str:
    """Generate a short session title (≤10 chars) via a non-streaming call."""
    opts = options or {}
    model: str = opts.get("model", "")
    base_url: str | None = opts.get("baseURL")

    client = _create_client(api_key, base_url)

    res = await client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": SESSION_TITLE_SYSTEM_PROMPT},
            {"role": "user", "content": str(text or "").strip()},
        ],
        max_tokens=32,
        extra_body={"thinking": {"type": "disabled"}},
        stream=False,
    )
    raw = (res.choices[0].message.content if res.choices and res.choices[0].message else "") or ""
    logger.info("[ai-generate-title][openai] 模型返回原文: %s", raw)
    return normalize_session_title(raw)
